File: archive/speedgrader/app/src/main/java/com/instructure/speedgrader/dialogs/DatePickerFragment.java
/*
 * Copyright (C) 2016 - present  Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.instructure.speedgrader.dialogs;

import android.app.DatePickerDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.widget.DatePicker;

import com.instructure.speedgrader.views.DateTextView;

import java.util.Calendar;
import java.util.Date;

public class DatePickerFragment extends DialogFragment
                            implements DatePickerDialog.OnDateSetListener {

    public final static String tag = "date_picker_tag";

    boolean wasCancelled;
    boolean shouldModifyTime;
    private Calendar calendar;

    public static DatePickerFragment getInstance(DateTextView dateTextView, boolean shouldModifyTime){

        DatePickerFragment datePickerFragment = new DatePickerFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable("dtv", dateTextView);
        bundle.putBoolean("time", shouldModifyTime);
        datePickerFragment.setArguments(bundle);

        return datePickerFragment;
    }

    private DateTextView dateTextView;

    //We need to prevent duplicates. If the user leaves the page, we don't want two DatePickerFragments.
    @Override
    public void onPause(){
        super.onPause();
        wasCancelled = true;
        dismiss();
    }


    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        // Use the current date as the default date in the picker
        shouldModifyTime = getArguments().getBoolean("time",false);
        dateTextView = (DateTextView) getArguments().getSerializable("dtv");
        Date date = dateTextView.getDate();

        //Error Checking
        if(date == null){
            date = new Date();
        }

        //DatePickerDialog expects data in a way that calendar provides natively.
        calendar = Calendar.getInstance();
        calendar.setTime(date);

        // Create a new instance of DatePickerDialog and return it
        return new DatePickerDialog(getActivity(), this, calendar.get(Calendar.YEAR),  calendar.get(Calendar.MONTH), calendar.get(Calendar.DATE));
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState){
        super.onActivityCreated(savedInstanceState);

        ((DatePickerDialog)getDialog()).getDatePicker().setCalendarViewShown(true);
        ((DatePickerDialog)getDialog()).getDatePicker().setSpinnersShown(false);

        ((DatePickerDialog)getDialog()).setOnCancelListener(new DialogInterface.OnCancelListener() {
            @Override
            public void onCancel(DialogInterface dialogInterface) {
                wasCancelled = true;
            }
        });

        ((DatePickerDialog)getDialog()).setOnDismissListener(new DialogInterface.OnDismissListener() {
            @Override
            public void onDismiss(DialogInterface dialogInterface) {
                if(!wasCancelled){
                    dateTextView.setDate(calendar.getTime());

                    if (shouldModifyTime) {
                        TimePickerFragment timePickerFragment = TimePickerFragment.getInstance(dateTextView);
                        timePickerFragment.show(getFragmentManager(), "time_picker");
                    }
                }



            }
        });

    }

    public void onDateSet(DatePicker view, int year, int month, int day) {
        // Convert the data back to how the dateTextView expects.
            calendar.set(year, month, day);
    }
}

File: apps/student/src/main/java/com/instructure/student/fragment/DatePickerFragment.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.instructure.student.fragment

import android.app.DatePickerDialog
import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.widget.DatePicker
import androidx.fragment.app.DialogFragment
import java.util.*

class DatePickerFragment : DialogFragment(), DatePickerDialog.OnDateSetListener, DialogInterface.OnCancelListener {

    var datePickerListener: DatePickerFragmentListener? = null
    var cancelListener: DatePickerCancelListener? = null

    private var year = -1
    private var month = -1
    private var day = -1

    interface DatePickerFragmentListener {
        fun onDateSet(year: Int, month: Int, day: Int)

    }

    interface DatePickerCancelListener {
        fun onCancel()

    }

    fun notifyDatePickerCancelListener() {
        if (cancelListener != null) {
            cancelListener!!.onCancel()
        }
    }

    fun notifyDatePickerListener(year: Int, month: Int, day: Int) {
        if (datePickerListener != null) {
            datePickerListener!!.onDateSet(year, month, day)
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        //use the current day as the default
        val c = Calendar.getInstance()
        if (year != -1 && month != -1 && day != -1) {
            return DatePickerDialog(requireContext(), this, year, month, day)
        } else {
            return DatePickerDialog(requireContext(), this,
                    c.get(Calendar.YEAR),
                    c.get(Calendar.MONTH),
                    c.get(Calendar.DAY_OF_MONTH))
        }
    }

    override fun onDateSet(datePicker: DatePicker, year: Int, month: Int, day: Int) {
        // Here we call the listener and pass the info back to it.
        notifyDatePickerListener(year, month, day)
    }

    override fun onCancel(dialog: DialogInterface) {
        super.onCancel(dialog)
        notifyDatePickerCancelListener()
    }

    companion object {

        fun newInstance(listener: DatePickerFragmentListener, cancelListener: DatePickerCancelListener): DatePickerFragment {
            val fragment = DatePickerFragment()
            fragment.datePickerListener = listener
            fragment.cancelListener = cancelListener
            return fragment
        }

        fun newInstance(listener: DatePickerFragmentListener, cancelListener: DatePickerCancelListener, year: Int, month: Int, day: Int): DatePickerFragment {
            val fragment = DatePickerFragment()
            fragment.datePickerListener = listener
            fragment.cancelListener = cancelListener
            fragment.year = year
            fragment.month = month
            fragment.day = day
            return fragment
        }
    }
}
